internal/k8s/k8s.go

```diff
@@ -358,6 +358,7 @@ func (c *Client) sync(key interface{}) SyncState {
 				return SyncStateError
 			}
 			if !exists {
+				l.Log("op", "getEndpoints", "error", err, "msg", "no endpoints defined")
```

Comment:
`err` is nil in this case, right?

Have you run this in your setup? What is the output it shows?

Reply:
Yes err is nil. 

This is the log line that it provides.
`{"caller":"k8s.go:361","error":null,"msg":"no endpoints defined","op":"getEndpoints","service":"default/registry","ts":"2020-03-29T16:11:00.865053157Z"}`

Reply:
Why print the error when there is no error? That seems buggy to me :)

Reply:
Now I understand why you were asking about err being nil. I can walk you through my thought process. I'm not really sure what else it could/should be. 

Even though metallb doesn't use log levels to determine how verbose logging should be they are typically ALL, TRACE, DEBUG, INFO, WARN, ERROR, FATAL

I can eliminate trace, debug, info, and fatal right off the bat as they just don't make sense for an error message that is relevant to normal operation. 

Warn vs Error could both be reasonable options. When I think of warning it's something that could be wrong but probably doesn't affect normal operation. Error is an indication of a problem that should be dealt with. It's not going to crash but probably isn't going to work the way you expect it to. 

Error just feels like the right answer to me. I'm expecting metallb to respond to the arp requests for the service I created. However under this condition it's not going to do that. 

If not error what do you think would be a more appropriate way to log the information?

Thanks,

James

Reply:
Sorry I wasn't clear, wasn't talking about that. I mean why print `error: null`. We *know* it is nil, there is no point in printing.

In particular, this part of the code `"error", err,` is what I'm talking about

Reply:
Ah ok. That did not register for me at all. I think having the word error in there is important as it will draw users to it. The logging is essentially key/value pairs so I don't think we can have nothing there so 2 potential options come to mind.
1. I could write some static text there. e.g error, "No endpoints defined" and drop the msg. 
2. I could write some static text there e.g error, "No edmpoints defined" and add a different message with more detail msg, "run kubectl get endpoints to validate service endpoints"

What do you think?

Reply:
I think a short description and msg seems reasonable :)

Reply:
Ok. I've made another commit for the change. The log now looks like this.
{"caller":"k8s.go:361","error":"No endpoints found","msg":"run kubectl get endpoints to validate that service endpoints exist","op":"getEndpoints","service":"default/registry","ts":"2020-04-01T15:06:47.810393689Z"}

Reply:
Sorry, thinking about it a little more, what about as msg: `No endpoints for service found`?

Reply:
@Matoch maybe also talk about verifying labels and selectors ?

Reply:
@rata Ok. I adjusted the wording again. It will now say error:No endpoints for service found and msg:run kubectl get endpoints to review service endpoints for service.

@champtar I can add that but I wasn't sure of all the possible causes for why endpoints may not exist. My thought was to give them enough info to know that the issue exists outside of metallb and then they can troubleshoot on their own. 

For both of you. I don't do pull requests often. If I read this correctly my additional commits do exist in the same pull request. Can you confirm if the updated wording is showing up for you?

Reply:
I'd not give troubleshooting or debugging instructions here. I'd just point that no endpoints are available, with the message I suggested: `No endpoints for service found`.

And I'm on the fence if we should do this (as this is really k8s info, not metallb), but I guess we can try to be helpful.

Then, in the troubleshooting guides we can explain more (it can be pods not ready, selector mismatch, etc.) if we want to. But, this is really a kubernetes thing, not very related to metallb, so I think I'd link to this: https://kubernetes.io/docs/tasks/debug-application-cluster/debug-service/

Reply:
@rata I'm certainly fine with keeping it simple. For me I finally figured out my issue as soon as I figured out that there were no endpoints. That gave me enough info to go look up what I needed to verify there was an issue with my service and move forward.

Reply:
I looked at it today. I found two concerns for now:

 * The patch doesn't work as intended: created a simple service with selector not matching anything and the line is not printed. I used `inv dev-env`, see https://metallb.universe.tf/community/#required-software for software needed to run. I tried using a configmap with just a layer2 IP address
 * It seems that the logic is added in the generic watcher code (shared by controller and speaker) and that the cache key doesn't exist doesn't mean that there are no ready endpoints

Can you please elaborate, so I can reproduce myself, how are you testing this? I might be missing something (starting by bgp/layer2 mode you are using, etc.)

The `ShouldAnnounce()` and it's caller might be a better candidate for this. The function already returns with a reason, and it is aware of the endpoints for the service. Although the reason is not logged (it might be too verbose) now.

Maybe the shared watcher code where you added it can be a better place, I'm not very familiar with it yet, but the code doesn't seem to work at all for a simple test.

Am I missing something?

Reply:
These are the steps I followed to test the code.

cd metallb/speaker
GOOS=linux GOARCH=amd64 go build
cp speaker ~/tmp/metallb-speaker/
cd ~/tmp/metallb-speaker

cat Dockerfile 
FROM metallb/speaker:v0.8.2
COPY speaker /

docker build -t matoch/metallb-speaker2 . && docker push matoch/metallb-speaker2
ssh to where I can run kubectl commands.

wget https://raw.githubusercontent.com/google/metallb/v0.9.3/manifests/metallb.yaml
# replace speaker image with matoch/metallb-speaker2
kubectl apply -f metallb.yaml

# My registry deployment uses the labels run: registry
# My config for the registry loadbalancer has a match label of app: registry just so that the match fails.
# After metallb is running and I load the registries load balancer config I get the following in my logs

{"caller":"k8s.go:361","error":"No endpoints for service found","msg":"run kubectl get endpoints to review service endpoints for service","op":"getEndpoints","service":"default/registry","ts":"2020-04-05T21:01:30.298333756Z"}


Now I'm not sure if the order of operations matters. I generally teardown my registry configs before redeploying metallb with the custom speaker.

 I'll take a look at the inv dev-env. I've only tested it on my cluster.

James

Reply:
Thanks a lot for the clarification! However, testing it again, such a message will be shown for all services without endpoint, not just services with type load balancer. 

So, if you create a service type clusterip it will show a message. Probably we don't want that, and not at that layer of abstraction. As I mentioned, that is the sync code shared for speaker and controller, that is why you probably see that entry before the `start of service update` one.

I'll paste another two lines diff, but in the general discussion (this is becoming too long and difficult to follow :))